File: roblox/Encoder.lua
if not game:GetService("RunService"):IsStudio() then return nil end
if game:GetAttribute("HarnessXEnabled") ~= true then return nil end

-- Studio-only transport encoder.
--
-- This is reversible transport obfuscation, not encryption. It applies a
-- rolling XOR using a timestamp-derived nonce, then writes the transformed
-- bytes as hexadecimal. server/encoder.py implements the matching decoder.

local HttpService = game:GetService("HttpService")

local Encoder = {}
local UINT32 = 4294967296

local function keyByte(nonceByte: number, roundIndex: number, byteIndex: number): number
	return (nonceByte + roundIndex * 31 + byteIndex * 17) % 256
end

local function transform(input: string, nonceByte: number, rounds: number): string
	local output = input

	for roundIndex = 1, rounds do
		local transformed = table.create(#output)

		for byteIndex = 1, #output do
			local sourceByte = string.byte(output, byteIndex)
			local key = keyByte(nonceByte, roundIndex, byteIndex)
			transformed[byteIndex] = string.char(bit32.bxor(sourceByte, key))
		end

		output = table.concat(transformed)
	end

	return output
end

local function toHex(input: string): string
	local output = table.create(#input)

	for index = 1, #input do
		output[index] = string.format("%02x", string.byte(input, index))
	end

	return table.concat(output)
end

local function fromHex(input: string): (string?, string?)
	if #input % 2 ~= 0 or string.match(input, "^[0-9a-fA-F]*$") == nil then
		return nil, "Invalid hexadecimal payload"
	end

	local output = table.create(#input / 2)
	local outputIndex = 0

	for index = 1, #input, 2 do
		outputIndex += 1
		local byteValue = tonumber(string.sub(input, index, index + 1), 16)
		if byteValue == nil then
			return nil, "Invalid hexadecimal byte"
		end
		output[outputIndex] = string.char(byteValue)
	end

	return table.concat(output), nil
end

function Encoder.encode(value: any, rounds: number): string
	assert(typeof(rounds) == "number", "rounds must be a number")
	rounds = math.floor(rounds)
	assert(rounds >= 1 and rounds <= 16, "rounds must be between 1 and 16")

	local nonce = DateTime.now().UnixTimestampMillis
	local nonceByte = nonce % 256
	local jsonPayload = HttpService:JSONEncode(value)
	local transformed = transform(jsonPayload, nonceByte, rounds)

	return HttpService:JSONEncode({
		version = 1,
		algorithm = "rolling-xor-hex",
		nonce = nonce,
		rounds = rounds,
		data = toHex(transformed),
	})
end

function Encoder.decode(envelopeJson: string): (any?, string?)
	local envelopeOk, envelope = pcall(function()
		return HttpService:JSONDecode(envelopeJson)
	end)

	if not envelopeOk or typeof(envelope) ~= "table" then
		return nil, "Envelope is not valid JSON"
	end

	if envelope.version ~= 1 or envelope.algorithm ~= "rolling-xor-hex" then
		return nil, "Unsupported encoder envelope"
	end

	if typeof(envelope.nonce) ~= "number"
		or typeof(envelope.rounds) ~= "number"
		or typeof(envelope.data) ~= "string"
	then
		return nil, "Envelope fields are invalid"
	end

	local transformed, hexError = fromHex(envelope.data)
	if transformed == nil then
		return nil, hexError
	end

	-- XOR is self-inverse, but multi-round decoding must reverse round order.
	local output = transformed
	local nonceByte = envelope.nonce % 256

	for roundIndex = math.floor(envelope.rounds), 1, -1 do
		local decoded = table.create(#output)

		for byteIndex = 1, #output do
			local sourceByte = string.byte(output, byteIndex)
			local key = keyByte(nonceByte, roundIndex, byteIndex)
			decoded[byteIndex] = string.char(bit32.bxor(sourceByte, key))
		end

		output = table.concat(decoded)
	end

	local payloadOk, payload = pcall(function()
		return HttpService:JSONDecode(output)
	end)

	if not payloadOk then
		return nil, "Decoded payload is not valid JSON"
	end

	return payload, nil
end

return table.freeze(Encoder)

File: server/encoder.py
from __future__ import annotations

import json
import time
from typing import Any

MAX_ROUNDS = 16


class EncoderError(ValueError):
    """Raised when a transport envelope cannot be decoded safely."""


def _key_byte(nonce_byte: int, round_index: int, byte_index: int) -> int:
    # Luau indexes bytes from 1, so Python callers pass byte_index + 1.
    return (nonce_byte + round_index * 31 + byte_index * 17) % 256


def _transform(data: bytes, nonce_byte: int, rounds: int, *, decode: bool) -> bytes:
    output = data
    round_order = range(rounds, 0, -1) if decode else range(1, rounds + 1)

    for round_index in round_order:
        output = bytes(
            byte ^ _key_byte(nonce_byte, round_index, byte_index + 1)
            for byte_index, byte in enumerate(output)
        )

    return output


def encode(value: Any, rounds: int = 3, *, nonce_ms: int | None = None) -> str:
    if not isinstance(rounds, int) or not 1 <= rounds <= MAX_ROUNDS:
        raise EncoderError(f"rounds must be an integer from 1 to {MAX_ROUNDS}")

    nonce = int(time.time() * 1000) if nonce_ms is None else int(nonce_ms)
    payload = json.dumps(
        value,
        ensure_ascii=False,
        separators=(",", ":"),
    ).encode("utf-8")
    transformed = _transform(payload, nonce % 256, rounds, decode=False)

    return json.dumps(
        {
            "version": 1,
            "algorithm": "rolling-xor-hex",
            "nonce": nonce,
            "rounds": rounds,
            "data": transformed.hex(),
        },
        separators=(",", ":"),
    )


def decode(envelope_text: str) -> Any:
    try:
        envelope = json.loads(envelope_text)
    except json.JSONDecodeError as exc:
        raise EncoderError("Envelope is not valid JSON") from exc

    if not isinstance(envelope, dict):
        raise EncoderError("Envelope must be a JSON object")
    if envelope.get("version") != 1:
        raise EncoderError("Unsupported envelope version")
    if envelope.get("algorithm") != "rolling-xor-hex":
        raise EncoderError("Unsupported envelope algorithm")

    nonce = envelope.get("nonce")
    rounds = envelope.get("rounds")
    data_hex = envelope.get("data")

    if not isinstance(nonce, int):
        raise EncoderError("nonce must be an integer")
    if not isinstance(rounds, int) or not 1 <= rounds <= MAX_ROUNDS:
        raise EncoderError(f"rounds must be an integer from 1 to {MAX_ROUNDS}")
    if not isinstance(data_hex, str):
        raise EncoderError("data must be a hexadecimal string")
    if len(data_hex) % 2 != 0:
        raise EncoderError("hexadecimal data has an odd length")

    try:
        transformed = bytes.fromhex(data_hex)
    except ValueError as exc:
        raise EncoderError("data is not valid hexadecimal") from exc

    decoded = _transform(transformed, nonce % 256, rounds, decode=True)

    try:
        return json.loads(decoded.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise EncoderError("Decoded data is not valid UTF-8 JSON") from exc
